src/collider/BVHTree.cpp

```diff
@@ -322,91 +223,138 @@ uint32_t SingleBVHTree::Partition(SortKey *keys, uint32_t numKeys, const AABBd &
-		}
+	AABBd aabb = AABBd::Invalid();
+	m_treeHeight = std::max(++height, m_treeHeight);
 
-		// Push in reverse order for pre-order traversal
-		stack[stackLevel++] = node->kids[1];
-		stack[stackLevel++] = node->kids[0];
+	// Compute the AABB for this node
+	for (size_t idx = 0; idx < numKeys; idx++) {
+		aabb.Update(objAabbs[keys[idx].index]);
 	}
-}
 
-void SingleBVHTree::TraceRay(const vector3d &start, const vector3d &inv_dir, double len, std::vector<uint32_t> &out_isect) const
-{
-	PROFILE_SCOPED()
+	node->aabb = aabb;
+	node->leafIndex = 0;
+	node->treeHeight = height;
 
-	int32_t stackLevel = 0;
-	uint32_t *stack = stackalloc(uint32_t, m_treeHeight + 1);
-	stack[stackLevel++] = 0;
+	// With only a single item, this is a leaf node. Set both child pointers to 0
+	if (numKeys == 1) {
+		node->kids[0] = node->kids[1] = 0;
+		node->leafIndex = keys[0].index;
+		return;
+	}
 
-	while (stackLevel > 0) {
-		uint32_t nodeIdx = stack[--stackLevel];
-		const SingleBVHTree::Node *node = &m_nodes[nodeIdx];
+	uint32_t startIdx = Partition(keys, numKeys, aabb, objAabbs);
 
-		// Didn't intersect with the node, ignore it
-		if (!node->aabb.IntersectsRay(start, inv_dir, len))
-			continue;
+	// If partitioning multiple nodes failed, just split the list down the middle
+	// This should never occur except in extreme degenerate cases
+	if (std::min(startIdx, numKeys - startIdx) == 0)
+		startIdx = numKeys >> 1;
 
-		// Leaf node - mark intersection and continue
-		if (node->kids[0] == 0) {
-			out_isect.push_back(node->leafIndex);
-			continue;
+	// Allocate both child nodes together for cache optimization
+	node->kids[0] = m_nodes.size();
+	Node *leftNode = &m_nodes.emplace_back();
+
+	node->kids[1] = m_nodes.size();
+	Node *rightNode = &m_nodes.emplace_back();
+
+	// Descend into left/right node trees
+	// This could potentially be made into a non-recursive stack-based algorithm
+	BuildNode(leftNode, keys, startIdx, objAabbs, height);
+	BuildNode(rightNode, keys + startIdx, numKeys - startIdx, objAabbs, height);
+}
+
+uint32_t BinnedAreaBVHTree::Partition(SortKey *keys, uint32_t numKeys, const AABBd &aabb, const AABBd *objAabbs)
+{
+	uint32_t axis = 0;
+	float pivot = 0.0;
+	float cost = FLT_MAX;
+
+	for (uint32_t i = 0; i < 3; i++) {
+		float newCost = FLT_MAX;
+		float newPivot = FindPivot(keys, numKeys, aabb, objAabbs, i, newCost);
+		if (newCost < cost) {
+			axis = i;
+			pivot = newPivot;
+			cost = newCost;
 		}
+	}
 
-		stack[stackLevel++] = node->kids[1];
-		stack[stackLevel++] = node->kids[0];
+	// Simple O(n) sort algorithm to sort all objects according to side of pivot
+	uint32_t startIdx = 0;
+	uint32_t endIdx = numKeys;
+
+	// It is ~10% faster to sort the indices than to sort the whole AABB array
+	// (cache hit rate vs. memory bandwidth)
+	// Sorting in general is extremely fast.
+	while (startIdx < endIdx) {
+		if (keys[startIdx].center[axis] < pivot) {
+			startIdx++;
+		} else {
+			endIdx--;
+			std::swap(keys[startIdx], keys[endIdx]);
```

Comment:
This swap turns out to be useless if the element at the end is also greater.  Generally, if the array is already sorted, it seems to make a lot of unnecessary swaps. Maybe it's possible to check the element in `endIdx` before swap? (applies to all places where this algorithm is used)

Reply:
This broadly breaks down into two categories. Either the array is "unknown" - i.e. we cannot assume the elements are sorted - or the array is already known to be sorted.

1. If the array is unknown, we cannot determine whether the element at the end that we are swapping with is greater until we read it. Either we read it before swapping and conditionally swap based on the value, or we always swap and then read it as part of the next iteration (at which point the value can be cached in a register).
2. If the array is known to be sorted, this function will perform unnecessary swaps for approximately 1/2 of the array, amortized.

This function is used almost exclusively over unsorted data, as the pivot axis and point change with every successive call to `Partition()`. Because this does not perform an in-order ascending/descending sort, but rather a simple partition over the dataset, any change to pivot axis or point invalidates the sorted state of the array.

The algorithm is intentionally designed to be as simple and unconditional as possible. There is a single less-than check - the minimum possible to implement the algorithm - and the algorithm is intended to perform as few operations as possible to make the best use of instruction and data caches.

Reply:
I've just tested this to confirm my gut is correct - changing `SingleBVHTree::Partition` (as it's actually called per-frame rather than only once at load time) to the following:
```c++
	while (startIdx < endIdx) {
		if (keys[startIdx].center[axis] < pivot) {
			startIdx++;
		} else {
			do { endIdx--; } while (endIdx > startIdx && keys[endIdx].center[axis] > pivot);
			std::swap(keys[startIdx], keys[endIdx]);
		}
	}
```
Results in 2-4us slower execution time (15us -> 17-19us in a test scene).